pyquil/paulis.py

```diff
@@ -33,6 +33,9 @@
 from collections.abc import Sequence
 import warnings
 
+NumberT = Union[int, float, complex]
```

Comment:
What's the reason for making your own number type? I see that you still use `Number` in at least one case below. If this is truly necessary, we might consider using or aliasing the `ExpressionValueDesignator` type from @appleby's PR #999.

Reply:
This was discussed at https://github.com/rigetti/pyquil/pull/1075#issuecomment-546128754

Reply:
I will rename to `ExpressionValueDesignator`.

Reply:
I have not actually read over this PR yet, so apologies if my comments are over general or do not apply.

Regarding `Number` vs `Union[int, float, complex]`, I agree with [@rht's comment here](https://github.com/rigetti/pyquil/pull/1075#issuecomment-546613935), namely that the type hints should match whatever the code actually accepts via `isinstance` checks.

Assuming we want the type hints to match what the code actually does, I think that means we have two choices, neither of which is very appealing:

1. Leave all the type hints as specifying `Number` and accept that they are virtually useless for static checking via mypy, unless or until https://github.com/python/mypy/issues/3186 is addressed upstream. The type hints still serve some purpose by way of documentation. Also, I haven't spent very long thinking about the exact implications here, or whether there is some workaround that might work for us, but if `n : Number = 5` doesn't typecheck, it seems unlikely we'll be able to do much about it except sprinkle `# type: ignore` comments around.

2. Change the type hints and the `isinstance` checks in code (and any other code updates that might be required) to `Union[int, float, complex]` rather than `Number` everywhere. This would mean things can type check via mypy, but would be a backwards incompatible change for pyquil. For example, it looks like numpy numeric types are `Number`s, but not necessarily a subtype of the corresponding python type

```
>>> isinstance(np.int64(0), numbers.Number)
True
>>> isinstance(np.int64(0), int)
False
```

We could always extend the `Union` to include numpy types we care about, but it still might break things in the unlikely event someone is passing a `Decimal` or `Fraction` or `MyCustomNumericType`. I defer to @karalekas about wether such a breaking change would be acceptable.

We could always opt for (1) now, then file an issue to remind us to tackle (2) whenever we ship the next major release of pyquil with breaking changes allowed.

Reply:
@appleby seeing as these are just types, I feel like it's a pretty low-risk change to just go to the `Union` -- thoughts?

Reply:
We could change just the type hints without changing the corresponding `isinstance` checks and it (probably) won't break any user code, but (1) the type hints are now lying about what types the function actually accepts, and (2) it will still likely result in mypy typechecking errors in some cases because mypy uses `isinstance` checks in the code to help it deduce types. For example, suppose `MyClass.foo` is a method that returns a `float`. Then the following function fails to typecheck with mypy complaining `Item "float" of "Union[float, MyClass]" has no attribute "foo"` since it can't figure out that `float` is a `Number` and therefore will never have `.foo()` called on it. Note that this case is analogous to functions here that accept `Union[PauliSum, PauliTerm, int, float, complex]`.

```python
def bar(x: Union[float, MyClass]) -> float:
    if isinstance(x, Number):
        return x
    return x.foo()
```

In contrast, if you change the `isinstance` check to `float`, the following does typecheck succesfully

```python
def bar(x: Union[float, MyClass]) -> float:
    if isinstance(x, float):
        return x
    return x.foo()
```

In this case, you can avoid the error by inverting the `isinstance` check to instead check for `MyClass`, but the point remains that if the type hints don't match the `isinstance` checks, you might still end up confusing mypy.

```python
def bar(x: Union[float, MyClass]) -> float:
    if isinstance(x, MyClass):
        return x.foo()
    return x
``` 

Maybe it turns out all the cases in this file can be resolved by some similar simple trick, idk.

Reply:
I guess one thing we could try is changing every `isinstance(foo, Number)` to instead check `isinstance(foo, (Number, int, float, complex))`. It's a hack, but I *think* would give mypy enough help to correctly typecheck arguments declared `Union[int, float, complex]` while at the same time being backwards compatible.

I still don't love the idea that the type hints are definitely wrong, but maybe it's the lesser evil compared to getting spammed with mypy errors or breaking backwards compat.

Reply:
I tried the above suggestion out and it did reduce the number of type errors. For example, [this patch](https://gist.github.com/appleby/294bcd7075944fe4b0b443fc77442cd0) reduces the number of mypy errors when type checking this file from 70 to 35. At a glance, none of the remaining 35 look to be directly related to mypy being confused about how to type check `Number`.

Reply:
> I guess one thing we could try is changing every isinstance(foo, Number) to instead check isinstance(foo, (Number, int, float, complex)).

This is what I had in mind as a hack too. Additionally, `(Number, int, float, complex)` can be defined as `ExpressionValueClass` to avoid repetition.

Reply:
I would prefer something like `PauliNumericClass`, since it's otherwise confusing that `ExpressionValueDesignator` is `Union[int, float, complex]`, whereas `ExpressionValueClass` is `(Number, int, float, complex)`.  But otherwise, it seems like a reasonable compromise to me. If @karalekas doesn't object, I guess I'm OK with a small lie in the type hints in order to allow type checking to work.